File: ios/Application/CreateArchiveUseCase.swift
import Foundation

final class CreateArchiveUseCase {
  private let engine: ArchiveEngine

  init(engine: ArchiveEngine) {
    self.engine = engine
  }

  func execute(plan: CreationPlan, output: ArchiveOutput, onProgress: @escaping (ProgressSnapshot) -> Void) async throws -> CreationResult {
    try await engine.createArchive(plan: plan, output: output, onProgress: onProgress)
  }
}

File: android/src/main/java/com/mcodex/nitroarchive/application/CreateArchiveUseCase.kt
package com.mcodex.nitroarchive.application

import com.mcodex.nitroarchive.application.ports.*
import com.mcodex.nitroarchive.domain.ArchiveDomainError
import com.mcodex.nitroarchive.domain.ProgressSnapshot

class CreateArchiveUseCase(private val engine: ArchiveEngine) {

    suspend fun execute(
        plan: CreationPlan,
        output: ArchiveOutput,
        onProgress: (ProgressSnapshot) -> Unit = {}
    ): CreationResult {
        return engine.createArchive(plan = plan, output = output, onProgress = onProgress)
    }
}
